clib/cList.ml

```diff
@@ -160,20 +160,27 @@ let assign l n e =
 
 (** {6 Filtering} *)
 
-let rec filter_loop f p = function
-  | [] -> ()
-  | x :: l' as l ->
-    let b = f x in
-    filter_loop f p l';
-    if b then if p.tail == l' then p.tail <- l else p.tail <- x :: p.tail
+(* [filter_loop f (Some (c0,l0)) c l] will do c0.tail <- l0 if [for_all f l] *)
+let rec filter_loop f reset p = function
+  | [] -> begin match reset with
+      | None -> ()
+      | Some (c,orig) -> c.tail <- orig
+    end
+  | x :: l as orig ->
+    if f x then
+      let c = { head = x; tail = [] } in
+      let () = p.tail <- cast c in
+      filter_loop f (Option.append reset (Some (p,orig))) c l
```

Comment:
I don't like the use of the append combinator here. It makes the code less understandable (never remember which side takes precedence in case of Some / Some) and assuming no inlining it allocates 5 words at each iteration for no good reason.

Reply:
changed